pyquil/analytics.py

```diff
@@ -0,0 +1,310 @@
+    Returns the total number of qubits needed for this program
+    :param program: The program to analyze
+    :return: The total number of qubits allocated for this program
+    '''
+    return len(program.get_qubits())
+
+def get_gate_types(program):
+    '''
+    Returns a dictionary of gate types to their counts within the program
+    :param program: The program to analyze
+    :return: The mapping of gates to their counts within program
+    '''
+    counts = {}
+    for gate in get_all_gate_types():
+        counts[gate] = 0
+    
+    # Count how many gates there are for each gate type
+    for action in program.actions:
+        try:
+            counts[action[1].operator_name] += 1
+        except Exception:
+            try:
+                counts[action[1].op] += 1
+            except:
+                pass#print action[1]
+        
+    return counts
+
+def get_qubit_num_gates(gate_counts):
+    '''
+    Returns a dictionary with the original gate_counts, but with
+    extra entries 'Single Qubit Gates', 'Two Qubit Gates',
+    'Control Instructions', 'Classical Instructions',
+    'Three+ Qubit Gates', and 'Elements in DEFGATES'
+    '''
+    
+    results = deepcopy(gate_counts)
+    results['Single Qubit Gates'] = \
+                gate_counts['I'] + \
+                gate_counts['H'] + \
+                gate_counts['X'] + \
+                gate_counts['Y'] + \
+                gate_counts['Z'] + \
+                gate_counts['S'] + \
+                gate_counts['T'] + \
+                gate_counts['PHASE']
+    results['Two Qubit Gates'] = \
+                gate_counts['RX'] + \
+                gate_counts['RY'] + \
+                gate_counts['RZ'] + \
+                gate_counts['CNOT'] + \
+                gate_counts['CPHASE00'] + \
+                gate_counts['CPHASE01'] + \
+                gate_counts['CPHASE10'] + \
+                gate_counts['CPHASE'] + \
+                gate_counts['SWAP'] + \
+                gate_counts['ISWAP'] + \
+                gate_counts['PSWAP']
+    results['Control Instructions'] = \
+                gate_counts['RESET'] + \
+                gate_counts['NOP'] + \
+                gate_counts['WAIT'] + \
+                gate_counts['HALT']
+    results['Classical Instructions'] = \
+                gate_counts['TRUE'] + \
+                gate_counts['FALSE'] + \
+                gate_counts['NOT'] + \
+                gate_counts['AND'] + \
+                gate_counts['OR'] + \
+                gate_counts['MOVE'] + \
+                gate_counts['EXCHANGE'] + \
+                gate_counts['JUMP-WHEN'] + \
+                gate_counts['JUMP-UNLESS']
+    results['Three+ Qubit Gates'] = \
+                gate_counts['CCNOT'] + \
+                gate_counts['CSWAP']
+    results['Elements in DEFGATES'] = \
+                gate_counts['DEFGATE']
+    return results
+
+def analyze(program_creator, input_list, should_range=False, to_include=None, remove_zeros=False):
```

Comment:
boolean valued arguments (especially optional ones) are usually an example of trying to do too much in one function.  (here's a goofy story about that http://www.informit.com/articles/article.aspx?p=1392524)